File: sdk/typescript/src/managers/apikey.manager.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 KirkyX. All rights reserved.

import type {
  ApiKeyCreateOptions,
  ApiKeyCreateResponse,
  ApiKeyInfo,
  ApiKeyPermission,
  PaginationOptions,
  SuccessResponse,
  PaginationResult,
} from "../types/api.js";
import type { HttpClient } from "../utils/http.js";
import { SecureNotifyError } from "../types/errors.js";

/**
 * API key list result
 */
export interface ApiKeyListResult {
  keys: ApiKeyInfo[];
  pagination: PaginationResult;
}

/**
 * Options for listing API keys
 */
export interface ListApiKeysOptions extends PaginationOptions {
  userId?: string;
  isActive?: boolean;
}

/**
 * Revoke result
 */
export interface RevokeResult {
  success: boolean;
  deletedId: string;
  channelId?: string;
  deletedAt: string;
}

/**
 * API key manager for handling API key operations
 */
export class ApiKeyManager {
  private readonly http: HttpClient;
  private readonly basePath = "/api/keys";

  /**
   * Create a new API key manager
   */
  constructor(http: HttpClient) {
    this.http = http;
  }

  /**
   * Create a new API key
   *
   * @param options - API key creation options
   * @returns The created API key information (including the key)
   */
  async create(options: ApiKeyCreateOptions): Promise<ApiKeyCreateResponse> {
    if (!options.name) {
      throw SecureNotifyError.validation("name is required");
    }

    if (!options.userId) {
      throw SecureNotifyError.validation("userId is required");
    }

    if (!options.permissions || options.permissions.length === 0) {
      throw SecureNotifyError.validation("permissions are required");
    }

    const response = await this.http.post<SuccessResponse<ApiKeyCreateResponse>>(
      this.basePath,
      options
    );

    return response.data.data;
  }

  /**
   * Get API key information by ID
   *
   * @param id - The API key ID
   * @returns The API key information
   */
  async get(id: string): Promise<ApiKeyInfo> {
    if (!id) {
      throw SecureNotifyError.validation("id is required");
    }

    const response = await this.http.get<SuccessResponse<ApiKeyInfo>>(
      `${this.basePath}/${encodeURIComponent(id)}`
    );

    return response.data.data;
  }

  /**
   * List API keys with optional filtering and pagination
   *
   * @param options - Listing options
   * @returns The list of API keys with pagination info
   */
  async list(options?: ListApiKeysOptions): Promise<ApiKeyListResult> {
    const query: Record<string, string | number | boolean | undefined> = {};

    if (options?.limit !== undefined) {
      query['limit'] = options.limit;
    }
    if (options?.offset !== undefined) {
      query['offset'] = options.offset;
    }
    if (options?.userId !== undefined) {
      query['userId'] = options.userId;
    }
    if (options?.isActive !== undefined) {
      query['isActive'] = options.isActive;
    }

    const response = await this.http.get<SuccessResponse<ApiKeyInfo[]> & { pagination: PaginationResult }>(
      this.basePath,
      query
    );

    return {
      keys: response.data.data,
      pagination: response.data.pagination,
    };
  }

  /**
   * Revoke an API key
   *
   * @param id - The API key ID
   * @returns The revocation result
   */
  async revoke(id: string): Promise<RevokeResult> {
    if (!id) {
      throw SecureNotifyError.validation("id is required");
    }

    const response = await this.http.delete<SuccessResponse<RevokeResult>>(
      `${this.basePath}/${encodeURIComponent(id)}`
    );

    return response.data.data;
  }

  /**
   * Check if an API key is active
   *
   * @param id - The API key ID
   * @returns Whether the key is active
   */
  async isActive(id: string): Promise<boolean> {
    try {
      const key = await this.get(id);
      return key.isActive;
    } catch (error) {
      if (error instanceof SecureNotifyError && error.code === "NOT_FOUND") {
        return false;
      }
      throw error;
    }
  }

  /**
   * Get the permissions for an API key
   *
   * @param id - The API key ID
   * @returns The permissions array
   */
  async getPermissions(id: string): Promise<ApiKeyPermission[]> {
    const key = await this.get(id);
    return key.permissions;
  }

  /**
   * Check if an API key has a specific permission
   *
   * @param id - The API key ID
   * @param permission - The permission to check
   * @returns Whether the key has the permission
   */
  async hasPermission(id: string, permission: ApiKeyPermission): Promise<boolean> {
    const permissions = await this.getPermissions(id);
    return permissions.includes(permission);
  }
}

File: sdk/rust/src/managers/apikey_manager.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 KirkyX. All rights reserved.

//! API Key manager for SecureNotify SDK

use async_trait::async_trait;
use crate::{Result, SecureNotifyError};
use crate::types::api::*;

/// Trait for API key management operations
#[async_trait]
pub trait ApiKeyManager {
    /// Create a new API key
    async fn create_api_key(
        &self,
        name: &str,
        user_id: Option<&str>,
        permissions: Option<Vec<&str>>,
        expires_at: Option<&str>,
    ) -> Result<ApiKeyCreateResponse>;

    /// Get API key information
    async fn get_api_key(&self, key_id: &str) -> Result<ApiKeyInfo>;

    /// List all API keys
    async fn list_api_keys(
        &self,
        limit: Option<u32>,
        offset: Option<u32>,
    ) -> Result<Vec<ApiKeyInfo>>;

    /// Revoke an API key
    async fn revoke_api_key(&self, key_id: &str) -> Result<()>;
}

/// Implementation of ApiKeyManager
pub struct ApiKeyManagerImpl {
    http_client: std::sync::Arc<crate::utils::http::HttpClient>,
}

impl ApiKeyManagerImpl {
    /// Create a new ApiKeyManager
    pub fn new(http_client: std::sync::Arc<crate::utils::http::HttpClient>) -> Self {
        Self { http_client }
    }
}

#[async_trait]
impl ApiKeyManager for ApiKeyManagerImpl {
    async fn create_api_key(
        &self,
        name: &str,
        user_id: Option<&str>,
        permissions: Option<Vec<&str>>,
        expires_at: Option<&str>,
    ) -> Result<ApiKeyCreateResponse> {
        let request = ApiKeyCreateRequest {
            name: name.to_string(),
            user_id: user_id.map(|s| s.to_string()),
            permissions: permissions.map(|mut perms| perms.drain(..).map(|s| s.to_string()).collect()),
            expires_at: expires_at.map(|s| s.to_string()),
        };

        self.http_client.post("api/keys", &request).await.map_err(|e| e.into())
    }

    async fn get_api_key(&self, key_id: &str) -> Result<ApiKeyInfo> {
        let endpoint = format!("api/keys/{}", key_id);
        self.http_client.get(&endpoint).await.map_err(|e| e.into())
    }

    async fn list_api_keys(
        &self,
        limit: Option<u32>,
        offset: Option<u32>,
    ) -> Result<Vec<ApiKeyInfo>> {
        let mut endpoint = "api/keys".to_string();
        let mut params = Vec::new();

        if let Some(limit) = limit {
            params.push(format!("limit={}", limit));
        }
        if let Some(offset) = offset {
            params.push(format!("offset={}", offset));
        }

        if !params.is_empty() {
            endpoint.push('?');
            endpoint.push_str(&params.join("&"));
        }

        self.http_client.get(&endpoint).await.map_err(|e| e.into())
    }

    async fn revoke_api_key(&self, key_id: &str) -> Result<()> {
        let endpoint = format!("api/keys/{}/revoke", key_id);
        self.http_client.post_empty(&endpoint).await
    }
}
